### PR TITLE
Prevent crashes with malformed regex which match all lines of content

### PR DESCRIPTION
**Pull request by jgutman**
### Intent

Addresses #11348 part 1 -- prevent RStudio crashes

### Approach

So it turns out that system grep doesn't actually crash when you search for `()` or `[]` or other bad regexes, even though it doesn't really match any content. We give our Find in Files an escape hatch, where if the grep command has completed without error and returns actual output, but the matchOn/matchOff array is empty (this doesn't happen with true no results, where grep stdout is empty) we skip ahead to the next match. 

Essentially this gives us (No results found) when searching for regexes like `()`, but appropriate results when searching for `\(\)`. True no matches (ie searching for `shdjshfdsjgfsjkd`) continue to behave as usual, since the grep onStdOut callback is not triggered when grep finds no matches.

This is because grep doesn't error when you search for `()` -- it actually matches every line, every file, even blank lines.

Separately, we can put in a PR to allow for escaping these characters when in WholeWord mode

### Automated Tests

Existing unit tests pass. We can add a new automated test case in rstudio-ide-automation once this is closed

### QA Notes

Find in files  with regex selected:
Create a file
````
print()

hello()

[1-3]
````
1) `\(\)`
2) `()`
3) `nooooooo`
4) `[]`
5) `\[\]`

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ ] This PR passes all local unit tests

### CODE REVIEW

**Comment by kevinushey** [comment]
If I understand correctly, using "Whole word" causes us to treat the user text as a regular expression, even though (I think) the intention here is to treat the search string as-is. At least from the UI, the `[ ] Regular expression` checkbox remains unchecked, so that was my expectation.

I think that's happening on the client side here:

https://github.com/rstudio/rstudio/blob/9ab60a7b67570e0b0497b76afb8daab0b291421b/src/gwt/src/org/rstudio/studio/client/workbench/views/output/find/FindOutputPresenter.java#L507-L509

Do we know whether support for literal escapes in our regular expression search is supported? For example, we could instead use:

```
\b\Q<query>\E\b
```

to force the query to be treated as-is ... although technically, this would be a change in behavior, but I do think it better matches the user intention.

**Comment by kevinushey** [comment]
For git grep, this works:

```
git grep -P '\b\Q()\E\b'
```

And on macOS, this seems okay:

```
git grep -E '\b\Q()\E\b'
```

**Comment by jgutman** [comment]
That actually wasn't the full intention of this PR; I had a draft PR up that escapes characters that should be treated as literals in whole word mode; however we also noticed that searching for `()` in regex mode also crashes the session. Even though that's a bad regex, I don't think it should crash the session; this ensures that it won't lead to a crash. It doesn't fully fix the issue of allowing these regex characters to be treated as literals in whole word mode, which is what I was trying to do in the other PR.

File: src/cpp/session/modules/SessionFind.cpp [hunk 1]
```diff
@@ -1027,6 +1034,9 @@ class GrepOperation : public boost::enable_shared_from_this<GrepOperation>
             processContents(&lineInfo.decodedPreview, &lineInfo.decodedContents,
                &matchOn, &matchOff);
 
+            if (matchOn.getSize() == 0)
```
Comment (melissa-barca) [inline src/cpp/session/modules/SessionFind.cpp:1037]: Can we add a comment here about when this is the case?

**Review by melissa-barca — APPROVED** [review]
LGTM!

**Comment by kevinushey** [comment]
> That actually wasn't the full intention of this PR; I had a draft PR up that escapes characters that should be treated as literals in whole word mode; however we also noticed that searching for () in regex mode also crashes the session. Even though that's a bad regex, I don't think it should crash the session; this ensures that it won't lead to a crash. It doesn't fully fix the issue of allowing these regex characters to be treated as literals in whole word mode, which is what I was trying to do in the other PR.

Ahh, gotcha. Thank you for clarifying!

**Review by kevinushey — APPROVED** [review]

**Comment by jgutman** [comment]
@kevinushey Thanks for the suggestion! Unfortunately the `\Q()\E` syntax doesn't appear to work on the version of grep we bundle with Windows 😭 so I think I'll have to stick to \escaping \every \single \special \character 🙃

**Pull request merged**